File: epi_judge_java/epi/IsStringDecomposableIntoWords.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;
import epi.test_framework.TimedExecutor;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Set;

public class IsStringDecomposableIntoWords {

    public static List<String> decomposeIntoDictionaryWords(String domain, Set<String> dictionary) {
        int[] prev = new int[domain.length() + 1];
        Arrays.fill(prev, -1);

        if (isSuccess(domain, 0, prev, dictionary)) {
            List<String> result = new ArrayList<>();
            int curr = domain.length();
            do {
                int start = prev[curr];
                result.add(domain.substring(start, curr));
                curr = start;
            } while (prev[curr] != -1);

            Collections.reverse(result);
            return result;
        }

        return List.of();
    }

    private static boolean isSuccess(String domain, int start, int[] prev, Set<String> dictionary) {
        if (start == domain.length()) return true;

        for (int end = start; end < domain.length(); end++) {
            if (prev[end + 1] != -1) continue;
            if (dictionary.contains(domain.substring(start, end + 1))) {
                prev[end + 1] = start;
                if (isSuccess(domain, end + 1, prev, dictionary)) {
                    return true;
                }
            }
        }

        return false;
    }

    @EpiTest(testDataFile = "is_string_decomposable_into_words.tsv")
    public static void decomposeIntoDictionaryWordsWrapper(TimedExecutor executor,
                                                           String domain,
                                                           Set<String> dictionary,
                                                           Boolean decomposable)
            throws Exception {
        List<String> result =
                executor.run(() -> decomposeIntoDictionaryWords(domain, dictionary));

        if (!decomposable) {
            if (!result.isEmpty()) {
                throw new TestFailure("domain is not decomposable");
            }
            return;
        }

        if (result.stream().anyMatch(s -> !dictionary.contains(s))) {
            throw new TestFailure("Result uses words not in dictionary");
        }

        if (!String.join("", result).equals(domain)) {
            throw new TestFailure("Result is not composed into domain");
        }
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "IsStringDecomposableIntoWords.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/is_string_decomposable_into_words.cc
#include <string>
#include <unordered_set>
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"
using namespace std;

vector<string> DecomposeIntoDictionaryWords(
    const string &domain, const unordered_set<string> &dictionary) {
    vector<bool> dp(domain.size() + 1, false);
    vector<string> path(domain.size() + 1);
    dp[0] = true;
    path[0] = "";
    for (int end = 1; end <= domain.size(); end++) {
        for (const string &word: dictionary) {
            auto word_size = word.size();
            if (end >= word_size
                && dp[end - word_size]
                && domain.substr(end - word_size, word_size) == word) {

                path[end] = word;
                dp[end] = true;
            }
        }
    }

    vector<string> ans;
    if (dp[domain.size()]) {
        int idx = domain.size();
        while (idx > 0) {
            ans.push_back(path[idx]);
            idx -= path[idx].size();
        }
        reverse(ans.begin(), ans.end());
    }
    return ans;
}

void DecomposeIntoDictionaryWordsWrapper(
    TimedExecutor &executor, const string &domain,
    const unordered_set<string> &dictionary, bool decomposable) {
    vector<string> result = executor.Run(
        [&] { return DecomposeIntoDictionaryWords(domain, dictionary); });
    if (!decomposable) {
        if (!result.empty()) {
            throw TestFailure("domain is not decomposable");
        }
        return;
    }

    if (std::any_of(std::begin(result), std::end(result),
                    [&](const std::string &s) { return !dictionary.count(s); })) {
        throw TestFailure("Result uses words not in dictionary");
    }

    if (std::accumulate(std::begin(result), std::end(result), string()) !=
        domain) {
        throw TestFailure("Result is not composed into domain");
    }
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{
        "executor", "domain", "dictionary",
        "decomposable"
    };
    return GenericTestMain(args, "is_string_decomposable_into_words.cc",
                           "is_string_decomposable_into_words.tsv",
                           &DecomposeIntoDictionaryWordsWrapper,
                           DefaultComparator{}, param_names);
}
